NodeJS without $persist_dir

lukesampson opened the pull request:
How about this? It's an alternative NodeJS manifest using persistence but without requiring the `$persist_dir` variable in `env_add_path` or `post_install`.

This way, we can keep the complexities of persisted directories hidden from manifest authors and users (reducing the learning curve).

This change would also make it unnecessary for the workaround to check for `$` variables in `prepare_env_path`.

From a user's point of view, they would just need to know that the current version of NodeJS is installed at `~/scoop/apps/nodejs/current`, any path or other environment variables will reference this too, and the `bin` directory inside that `current` directory will be persisted across updates.

Perhaps this would also help us avoid having to add complicated source and target definitions in `persist` to move files, as you suggested might be necessary in future. I would much prefer to keep things simple, if possible. 

Ref: lukesampson/scoop#1372.

lukesampson committed:
A NodeJS install with persistence that doesn't use $persist_dir variable in manifest

lukesampson committed:
A NodeJS install with persistence that doesn't use $persist_dir variable in manifest

rrelmy commented on lib/install.ps1:
```diff
@@ -739,6 +739,7 @@ function link_current($versiondir) {
     }
 
     cmd /c mklink /j $currentdir $versiondir | out-null
+    attrib $currentdir +R /L
```
I think this is causing an error while uninstalling an app
![image](https://cloud.githubusercontent.com/assets/442683/24432151/71ba9d50-1420-11e7-82bd-685a8e6a0719.png)

"Zugriff verweigert" is german for "Permission denied" ;-)

rrelmy replied:
And it also breaks the relink on updates! After an update the current directory still points to the old version!

lukesampson committed:
Fix 'Access is denied' error removing 'current' directory junction.

lukesampson committed:
Fix 'Access is denied' error removing 'current' directory junction.

rrelmy committed:
persist nodejs fix npm updating itself

rrelmy committed:
persist nodejs fix npm updating itself

rrelmy commented:
This breaks npm updating itself!
npm can be run because the nodejs directory is in the path too. But because the prefix is changed npm does not find itself as a module and update it. Currently npm version 4.1.2 is bundled with nodejs but the latest version is 4.5.

We would need to move the npm stuff inside the bin directory to make that work. Therefore my move proposal.
Adding a `post_install` script like 7bfeb2aef8920016efee7f22493f3444afdb51b4 would fix it … but, eeeeeh :-)

lukesampson committed:
Unlink 'current' directory junction when updating app.

lukesampson committed:
Merge branch 'persist' of https://github.com/lukesampson/scoop into persist

lukesampson commented:
How are you updating NPM? Without 7bfeb2aef8920016efee7f22493f3444afdb51b4, I am able to run `npm install -g npm` and it works (although it only installs version 4.4). The version in `~/scoop/apps/nodejs/current/bin` takes precedence over the version in `~/scoop/apps/nodejs/current`, due to the order Scoop adds directories to the path.

Installing NodeJS from their official MSI installer seems to have the same behavior—i.e. running `npm update -g npm` does nothing, but `npm install -g npm` updates to version 4.4.

rrelmy commented:
Interesting that the official installer has the same behaviour.

What should we do?
1. Inherit the behaviour of the official installer and keep an old npm version
2. Keep the scoop (stable) behaviour with 7bfeb2aef8920016efee7f22493f3444afdb51b4 and have the latest (pre-release) version
3. Instead of 7bfeb2aef8920016efee7f22493f3444afdb51b4 run `npm install -g npm` inside post_install to receive the latest stable npm version and be able to keep it updated with `npm update -g` 

Personally I would prefer to have the latest (stable) version. But that could be achieved by running `npm install -g npm` manually once.

Version 4.4.4 is the current stable version, 4.5 is still pre-release. I am not sure why we get it after an update.

_Couldn't it be easy once? It always has to be complicated._ 😆

lukesampson commented:
Haha! Yes NodeJS is probably not the ideal test-case because it's so complicated, but I'm excited about this persistence feature overall. It looks like it will work, and will be a big improvement for Scoop, both in terms of "just working" for users, and making it easy for manifest authors to persist data.

I prefer option one, because:

* it keeps the manifest simple and easy to understand
* it's a quicker default install with nothing else to download
* it matches the official install (the perceived behavior and limitations), and is probably perfect for people who just want it to work
* people who do know about and want a later version of npm can opt-in by manually running the command as you said. Maybe they could even have upgraded npm before updating NodeJS.
* future official releases might come with a stable NPM (so other workarounds might be unnecessary)

rrelmy committed:
change nodejs install behaviour to match upstream installer

`npm update -g` has been remove because it has no effect on the
installed npm version and does not match the official installer
behaviour.

rrelmy commented:
Ok, I am ok with that.

I reverted the post_install script to match that behaviour, I also removed the `npm update -g` command.
It was introduced by 1896adb587acb13bc9840ae18bf0cbdc31208641 but does not have an effect on the installed npm version anymore.
